x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetJobModelSnapshotsUpgradeStatsAction.java

```diff
@@ -0,0 +1,128 @@
+
+    private static final Logger logger = LogManager.getLogger(TransportGetJobModelSnapshotsUpgradeStatsAction.class);
+
+    private final JobConfigProvider jobConfigProvider;
+
+    @Inject
+    public TransportGetJobModelSnapshotsUpgradeStatsAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        JobConfigProvider jobConfigProvider
+    ) {
+        super(
+            GetJobModelSnapshotsUpgradeStatsAction.NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            Request::new,
+            indexNameExpressionResolver,
+            Response::new,
+            ThreadPool.Names.SAME
+        );
+        this.jobConfigProvider = jobConfigProvider;
+    }
+
+    @Override
+    protected void masterOperation(Task task, Request request, ClusterState state, ActionListener<Response> listener) {
+        logger.debug(
+            () -> new ParameterizedMessage("[{}] get stats for model snapshot [{}] upgrades", request.getJobId(), request.getSnapshotId())
+        );
+        final PersistentTasksCustomMetadata tasksInProgress = state.getMetadata().custom(PersistentTasksCustomMetadata.TYPE);
+        final Collection<PersistentTasksCustomMetadata.PersistentTask<?>> snapshotUpgrades = MlTasks.snapshotUpgradeTasks(tasksInProgress);
+
+        // 2. Now that we have the job IDs, find the relevant model snapshot upgrades
+        ActionListener<List<Job.Builder>> expandIdsListener = ActionListener.wrap(jobs -> {
+            ExpandedIdsMatcher requiredSnapshotIdMatches = new ExpandedIdsMatcher(request.getSnapshotId(), request.allowNoMatch());
+            Set<String> jobIds = jobs.stream().map(Job.Builder::getId).collect(Collectors.toSet());
+            List<Response.JobModelSnapshotUpgradeStats> statsList = snapshotUpgrades.stream()
+                .filter(t -> jobIds.contains(((SnapshotUpgradeTaskParams) t.getParams()).getJobId()))
+                .filter(t -> requiredSnapshotIdMatches.idMatches(((SnapshotUpgradeTaskParams) t.getParams()).getSnapshotId()))
+                .map(t -> {
+                    SnapshotUpgradeTaskParams params = (SnapshotUpgradeTaskParams) t.getParams();
+                    Response.JobModelSnapshotUpgradeStats.Builder statsBuilder = Response.JobModelSnapshotUpgradeStats.builder(
+                        params.getJobId(),
+                        params.getSnapshotId()
+                    );
+                    return statsBuilder.setNode(state.getNodes().get(t.getExecutorNode()))
+                        .setUpgradeState(MlTasks.getSnapshotUpgradeState(t))
+                        .setAssignmentExplanation(t.getAssignment().getExplanation())
+                        .build();
```

Comment:
`t.getExecutorNode()` could be `null` if the task isn't yet assigned. We should short circuit this when `if (t.getExecutorNode() == null)`

Reply:
`state.getNodes().get(t.getExecutorNode())`

I am more worried that `state.getNodes().get(null)` throws